### PR TITLE
refactor(all): Repackaging

### PR DESCRIPTION
**Pull request by hansl**
Moving Angular-CLI into `packages/` and building the angular-cli.

The published packages will now have JavaScript only (with their d.ts), instead of being a mix of TypeScript and JavaScript, and it won't have the hacks of having a custom require extension to load the TypeScript.

Our development flow will be more streamlined, tools are going to be better (no more undefined symbols that are actually defined), and the npm packages will boot up way faster (since there's no more compilation during runtime).

Locally nothing changes, use `npm link` and run your local `ng`. Using the `npm` releases though it should be all javascript from the release.

The E2E tests also now uses the `dist` folder output to install in the new E2E app, making sure the process works for an app published to npm.

Note: Even though we're following the lerna structure, we won't be using their tool anytime soon.

### CODE REVIEW

File: packages/angular-cli/tsconfig.json [hunk 1]
```diff
@@ -25,9 +25,7 @@
     }
   },
   "exclude": [
-    "blueprints/*/files/**/*"
-  ],
-  "includes": [
-    "./custom-typings.d.ts"
+    "blueprints/*/files/**/*",
+    "**/*.spec.ts"
```
Comment (filipesilva) [inline packages/angular-cli/tsconfig.json]: What spec files are these?

File: packages/angular-cli/addon/index.js [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+  config: function () {
+    this.project.ngConfig = this.project.ngConfig || config.CliConfig.fromProject().config;
+  },
+
+  blueprintsPath: function () {
+    return path.join(__dirname, '../blueprints');
+  },
+
+  includedCommands: function () {
```
Comment (Brocco) [inline packages/angular-cli/addon/index.js]: Missing help command

File: plugins/karma.js [hunk 3]
```diff
@@ -1,43 +1 @@
...
-const preprocessor = () => (content, file, done) => done(null, content);
-preprocessor.$inject = []
-
-// also export karma-webpack and karma-sourcemap-loader
-module.exports = Object.assign({
-  'framework:angular-cli': ['factory', init],
-  'preprocessor:angular-cli': ['factory', preprocessor]
-}, require('karma-webpack'), require('karma-sourcemap-loader'));
\ No newline at end of file
+module.exports = require('../packages/angular-cli/plugins/karma');
```
Comment (filipesilva) [inline plugins/karma.js]: If you want you can just completely remove `plugin/karma.js` to something you think is a better import. It was only introduced in the webpack version so it's fine if you move it.

File: scripts/publish/build.js [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+    rs.on('error', reject);
+    ws.on('error', reject);
+    ws.on('close', resolve);
+
+    rs.pipe(ws);
+  });
+}
+
+
+function patch() {
```
Comment (filipesilva) [inline scripts/publish/build.js]: Can you add a comment explaining why this is necessary?

File: scripts/publish/build.js [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+      }, Promise.resolve());
+  })
+  .then(() => console.log('Copying uncompiled resources...'))
+  .then(() => glob(path.join(packagesRoot, '**/*')))
+  .then(files => {
+    console.log(`  Found ${files.length} files...`);
+    return files
+      .map((fileName) => path.relative(packagesRoot, fileName))
+      .filter((fileName) => {
+        if (/^angular-cli.blueprints/.test(fileName)) {
```
Comment (filipesilva) [inline scripts/publish/build.js]: Shouldn't this be `/^angular-cli\/blueprints/` instead? There doesn't seem to be any file with `angular-cli.blueprints` in it's name.

File: tsconfig.json [hunk 6]
```diff
@@ -6,7 +6,7 @@
     "moduleResolution": "node",
     "noEmitOnError": true,
     "noImplicitAny": true,
-    "outDir": "./dist/",
+    "outDir": "./dist/asd",
```
Comment (filipesilva) [inline tsconfig.json]: `asd`?

File: packages/angular-cli/tsconfig.json (hunk 1 above)
Comment (hansl) [inline packages/angular-cli/tsconfig.json]: We'll move acceptance specs into this package later on. I can remove if you want.

Comment (hansl) [inline packages/angular-cli/tsconfig.json]: I'm going to remove and put back. This tsconfig was copied from the one in ast-tools.

File: packages/angular-cli/addon/index.js (hunk 2 above)
Comment (hansl) [inline packages/angular-cli/addon/index.js]: oopsie. done.

File: plugins/karma.js (hunk 3 above)
Comment (hansl) [inline plugins/karma.js]: No I think it's a great import. This file is necessary when you `npm link` your local version. I've added a comment.

File: scripts/publish/build.js (hunk 4 above)
Comment (hansl) [inline scripts/publish/build.js]: It's not anymore, so removed.

File: scripts/publish/build.js (hunk 5 above)
Comment (hansl) [inline scripts/publish/build.js]: It's a regex, so `.` will match both `/` and `\`.

File: tsconfig.json (hunk 6 above)
Comment (hansl) [inline tsconfig.json]: Part of a test. You should never use this tsconfig, it's only there for tooling.

File: scripts/publish/build.js (hunk 5 above)
Comment (filipesilva) [inline scripts/publish/build.js]: That didn't even occur to me, but of course it makes sense.

**Comment by elvirdolic** [comment]
What is the goal/reason for this mega PR.

**Comment by hansl** [comment]
Clarified the text.

**Pull request merged**